### PR TITLE
Unify, simplify and fixup assignment errors

### PR DESCRIPTION
**Pull request by lydell**
- Show the same type of error message for compound assignment as for `=`
  assignment when the LHS is invalid.
- Show the same type of error message when trying to assign to a CoffeeScript
  keyword as when trying to assign to a JavaScript keyword.
- Now longer treat `&& =` as `&&=`. The same goes for `and=`, `||=` and `or=`.
- Unify the error message to: `<optional type> '<value>' can't be assigned`.

### CODE REVIEW

File: src/lexer.coffee [hunk 1]
```diff
@@ -411,14 +411,20 @@ exports.Lexer = class Lexer
...
+      skipToken = false
+      if value is '=' and prev[1] in ['||', '&&'] and not prev.spaced
         prev[0] = 'COMPOUND_ASSIGN'
         prev[1] += '='
-        return value.length
+        prev = @tokens[@tokens.length - 2]
+        skipToken = true
+      if prev and prev.variable
+        origin = prev.origin ? prev
+        if message = isUnassignable prev[1], origin[1]
```
Comment (michaelficarra) [inline src/lexer.coffee]: I find this pattern of doing an assignment in a condition really unreadable and error-prone.

File: src/lexer.coffee [hunk 2]
```diff
@@ -782,15 +803,12 @@ RESERVED = [
   'protected', 'public', 'static'
 ]
 
-STRICT_PROSCRIBED = ['arguments', 'eval', 'yield*']
+STRICT_PROSCRIBED = ['arguments', 'eval']
```
Comment (michaelficarra) [inline src/lexer.coffee]: This should include `yield`, shouldn't it?

Comment (lydell) [inline src/lexer.coffee]: But `yield` is a CoffeeScript keyword?

Comment (michaelficarra) [inline src/lexer.coffee]: Oh sure, if it's in that list then it shouldn't need to be in this list.

Comment (lydell) [inline src/lexer.coffee]: (It’s in `JS_KEYWORDS` (“keywords that CoffeeScript shares in common with JavaScript”).)

File: src/lexer.coffee (hunk 1 above)
Comment (lydell) [inline src/lexer.coffee]: Fixed all occurrences in the diff.

**Pull request merged**